### ISSUE TITLE
Sending requests blocks if there are not responders

### ISSUE THREAD

**Issue by Jarema** [nats-io/nats.rs#219]
#### NATS version (`grep 'name = "nats"' Cargo.lock -A 1`)
0.15.1
#### rustc version (`rustc --version` - we support Rust 1.41 and up)
1.51.0
#### OS/Container environment:
Ubuntu 21.04
#### Steps or code to reproduce the issue:
1. connect to NATS
2. send request to subject where are no responders
```
    let nc = nats::Options::new().with_name("republisher").connect("nats://localhost:4222")?;

    println!("publishing to stream");
    let result = nc.request("subjectwithoutresponders", "test");
    match result {
        Ok(ok) => println!("ok: {:?}", ok),
        Err(err) => println!("error: {}", err),
    }
    // nc.request will block. We will never reach this comment
    println!("done");
    ```
#### Expected result:
Get No Responders Error that is mapped from Status `503` from response headers
#### Actual result:
nc.request is blocking

**Comment by Jarema** [comment]
It seems that Rust client does not add `no_responders` property to CONNECT proto message:

```
/// Info to construct a CONNECT message.
#[derive(Clone, Debug)]
#[doc(hidden)]
#[allow(clippy::module_name_repetitions)]
pub struct ConnectInfo {
    /// Turns on +OK protocol acknowledgements.
    pub verbose: bool,

    /// Turns on additional strict format checking, e.g. for properly formed
    /// subjects.
    pub pedantic: bool,

    /// User's JWT.
    pub user_jwt: Option<SecureString>,

    /// Public nkey.
    pub nkey: Option<SecureString>,

    /// Signed nonce, encoded to Base64URL.
    pub signature: Option<SecureString>,

    /// Optional client name.
    pub name: Option<SecureString>,

    /// If set to `true`, the server (version 1.2.0+) will not send originating
    /// messages from this connection to its own subscriptions. Clients should
    /// set this to `true` only for server supporting this feature, which is
    /// when proto in the INFO protocol is set to at least 1.
    pub echo: bool,

    /// The implementation language of the client.
    pub lang: String,

    /// The version of the client.
    pub version: String,

    /// Indicates whether the client requires an SSL connection.
    pub tls_required: bool,

    /// Connection username (if `auth_required` is set)
    pub user: Option<SecureString>,

    /// Connection password (if auth_required is set)
    pub pass: Option<SecureString>,

    /// Client authorization token (if auth_required is set)
    pub auth_token: Option<SecureString>,

    /// Whether the client supports the usage of headers.
    pub headers: bool,
}
```

**Comment by derekcollison** [comment]
We will get that fixed up asap, thanks!

**Comment by Jarema** [comment]
@derekcollison I was able to pinpoint the issue:


Consider below NATS responses:
```
Ok("NATS/1.0\r\ntestkey: testval\r\nNats-Reply-Counter: 8\r\n\r\n") // proper response with headers
Ok("NATS/1.0 503\r\n\r\n") // 503 No Responders case
```

If second value is passed to 

```let headers = Headers::try_from(&*header_payload)?;```

here we have the problem:

```
let splits = line.splitn(2, ':').map(str::trim).collect::<Vec<_>>();
```

We do not have `:` delimetered key-values data in Headers in 503 case nor the code tries to match error response.

We can fix this by matching `NATS/1.0 503` in `try_from` and return new No Responders error there or straight after reading the paylaod in `decode`:

```
        // `HMSG <subject> <sid> [reply-to]
        // <# header bytes> <# total bytes>\r\n
        // <version line>\r\n[headers]\r\n\r\n[payload]\r\n`

        // Read the header payload.
        let mut header_payload = Vec::new();
        header_payload.resize(num_header_bytes as usize, 0_u8);
        stream.read_exact(&mut header_payload[..])?;

        // check for 503
        if  let Ok(s) = str::from_utf8(&header_payload) {
            if s.trim().eq("NATS/1.0 503") {
                // return No Responders here.
            }
        }

        let headers = Headers::try_from(&*header_payload)?;
```        

We could also map this to `Headers` and set the key for example as `status` with value `503` or `NoResponders` but that sounds a little like exposing NATS internal API to Client, unless we would check the Headers struct instance for 503 status before returning it. Though that is anyway an additional unnecessary instance of Headers Struct and in general it sounds like a bad idea (we would need to change `try_from` behaviour anyway and hardcode mathcing error responses there... bleh).

Sorry for any non-idiomatic code (Go dev trying Rust here)

I can push this topic and submit PR if we agree on solution.

**Comment by derekcollison** [comment]
Thanks, will have @spacejam take a look.
